### ISSUE TITLE
Jenkins start up failure because of plugin

### ISSUE THREAD

**Issue by akrasnov-drv** [jenkinsci/configuration-as-code-plugin#1291]
Today I had a weird issue when Jenkins could not start. I could see just Jenkins screen with Java stack trace.
- Jenkins version 2.204.2
- Plugin version 1.35 (last available within Jenkins)
- jenkins:lts-alpine image in Ubuntu 18 host

Appeared that during its start Jenkins tried to load the plugin, and the plugin tried to download configuration from the URL, that was used recently to update Jenkins. The URL was inaccessible (site was switched off), so plugin failed to load the configuration, and then Jenkins start was somehow broken (after 15 minutes it was at the same broken state). This issue may be  linked to "Config as code runs too early in the lifecycle #1026" but attempt to re-load last used config from the last used URL is quite bad issue itself.
Unfortunately java stack trace was not kept, and currently I have no option to reproduce it. 
When the site with config was brought up, and Jenkins restarted, the problem was solved. Without restart Jenkins kept printing java traces to log and did not come up. 
If it helps, the last config loaded by the plugin (and badly needed to start Jenkins) contained just EC2 fleets definitions.

**Comment by casz** [comment]
An issue with an workaround: https://github.com/jenkinsci/configuration-as-code-plugin#compatibility-with-jenkins--2199

and soon a proper fix: #1262

**Comment by akrasnov-drv** [comment]
@casz   I believe it's another issue than in the link you provided.

The problem I see - the plugin tries to download new configuration during Jenkins start. It tries to use URL that was used previously.

I believe it's both wrong behavior and security issue
The plugin should not bring any new configuration itself during Jenkins start.

My Jenkins start failure was just a follow up for the problem above.

**Comment by casz** [comment]
Best to set `CASC_JENKINS_CONFIG` as clearly stated in the readme.

**Comment by akrasnov-drv** [comment]
Ok, right, this may be considered as a workaround.

Still, I do not see any reason, why the plugin should update Jenkins configuration during Jenkins startup. You may add this feature as an option, but it should not be the default.
Maybe worth adding safe default to CASC_JENKINS_CONFIG (e.g. /dev/null)? This way you at least does not break something pre-existing...

**Comment by casz** [comment]
This is not a workaround this the intended way to configure CASC_JENKINS_CONFIG or otherwise default to what was configured previously.

Some people do not use containers and rely on what was previously persisted.

https://github.com/jenkinsci/configuration-as-code-plugin/blob/dcd748d6518204f44f7845a7afef80a8b6a601fe/plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java#L306-L337

**Comment by casz** [comment]
The reason why Jenkins should run configuration during startup is if you run a stateless Jenkins.

**Comment by casz** [comment]
If you consider this a security issue than do not point it to a file or an URL you do not trust. 🤷‍♂

**Comment by akrasnov-drv** [comment]
I see, but you describe just one of use cases. I believe there should be an option to choose it (e.g. by defining  CASC_JENKINS_CONFIG variable), but it should not be the default. And to remember the value provided in the GUI (server/configuration-as-code/) for the next Jenkins start is not correct at all.
I understand that the plugin was created in some scenario in mind, still I really think that it should be more "universal".
Anyway, I've got your point. Thank you for explanation. I'll try to define the var to some safe value in my system just to not fall into the "standard" workflow.

**Comment by casz** [comment]
You can always make your Jenkins partially stateless by using a different bind mount volume for only workspace/build as you said your running it on a container, thereby it will not "save" the global configuration for JCasC.

By setting `-Djenkins.model.Jenkins.workspacesDir` and `-Djenkins.model.Jenkins.buildsDir`

**Comment by casz** [comment]
Also what your describing above as the "universal" approach is what we got.

Some just use JCasC without any other configuration around Jenkins itself and rely on the global configuration path being saved. So when they restart JCasC it continues to work.

Other have committed to ansible, kubernetes, dockerfile to maintain their Jenkins instance and rely on setting JCASC_JENKINS_CONFIG.

Some have even gone all the way and do not persist anything on their Jenkins master and spin up several hundred of their Jenkins master a day.

I think we got the "universal" approach covered when we can cover all those bases.

Again the readme and the docs folder has everything detailed. Sometimes you cannot read everything.

**Comment by akrasnov-drv** [comment]
I see, and still our use case differs.
We use bind mount from docker host (configuration is kept).
We use the plugin to update EC2 cloud config (only).
We expected to update the configuration on request (and definitely not during Jenkins startup).
I just suggested some more acceptable (IMHO) behavior, if you think it's not interesting to follow my suggestion, just forget.
You provided me the input I needed to workaround the issue. Thanks.
